### PR TITLE
feat: Add option to merge model files at codegen time

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
https://github.com/awslabs/aws-sdk-swift/issues/1543

## Description of changes
Add an option to the smithy-swift codegen plugin that merges model files into a single `Models.swift` file.  This boosts compile speed when the number of model files is extremely high.  This option is off by default.

## Scope
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-swift-codegen-test/smithy-build.json [hunk 1]
```diff
@@ -10,7 +10,8 @@
                     "gitRepo": "https://github.com/smithy-lang/smithy-swift.git",
                     "author": "Amazon Web Services",
                     "homepage": "https://docs.amplify.aws/",
-                    "swiftVersion": "5.7.0"
+                    "swiftVersion": "5.7.0",
+                    "mergeModels": false
```
Comment (jbelkins) [inline smithy-swift-codegen-test/smithy-build.json:14]: This new option in `smithy-build.json` allows you to control merging of models.  I added it here even though it defaults to `false`.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/ServiceNamespaceIntegration.kt [hunk 2]
```diff
@@ -17,8 +18,8 @@ class ServiceNamespaceIntegration : SwiftIntegration {
         delegator: SwiftDelegator
     ) {
         val service = ctx.settings.getService(ctx.model)
-        val namespaceName = service.nestedNamespaceType(ctx.symbolProvider)
-        val filename = "${ctx.settings.moduleName}/models/$namespaceName.swift"
+        val namespaceName = service.nestedNamespaceType(ctx.symbolProvider).name
+        val filename = "${ctx.settings.moduleName}/${ModelFileUtils.filename(ctx.settings, namespaceName)}"
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/ServiceNamespaceIntegration.kt:22]: You will see several places in this project and in aws-sdk-swift where the `ModelFileUtils.filename()` method is used to resolve the model filename to be used.

I won't comment on all the changes similar to this one.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSettings.kt [hunk 3]
```diff
@@ -38,7 +39,8 @@ class SwiftSettings(
     val sdkId: String,
     val gitRepo: String,
     val swiftVersion: String,
-    var useInterceptors: Boolean
+    var useInterceptors: Boolean,
+    val mergeModels: Boolean,
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSettings.kt:43]: The `SwiftSettings` type (which is itself a property of generation context) is where you can find the merge models setting.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSettings.kt [hunk 4]
```diff
@@ -82,8 +85,21 @@ class SwiftSettings(
             val swiftVersion = config.expectStringMember(SWIFT_VERSION).value
             val sdkId = sanitizeSdkId(config.getStringMemberOrDefault(SDK_ID, serviceId.name))
             val useInterceptors = config.getBooleanMemberOrDefault(USE_INTERCEPTORS)
-
-            return SwiftSettings(serviceId, moduleName, version, desc, author, homepage, sdkId, gitRepo, swiftVersion, useInterceptors)
+            val mergeModels = config.getBooleanMemberOrDefault(MERGE_MODELS)
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSettings.kt:88]: This sets the `SwiftSettings` field based on the contents of `smithy-build.json`.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSymbolProvider.kt [hunk 5]
```diff
@@ -289,11 +290,8 @@ class SwiftSymbolProvider(private val model: Model, swiftSettings: SwiftSettings
     }
 
     private fun formatModuleName(shapeType: ShapeType, name: String): String? {
-        // All shapes except for the service are stored in models.
-        return when (shapeType) {
-            ShapeType.SERVICE -> "./$rootNamespace/${name}ClientProtocol.swift"
-            else -> "./$rootNamespace/models/$name.swift"
-        }
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftSymbolProvider.kt:296]: The support for service module name here is no longer used, and is removed.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/utils/ModelFileUtils.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package software.amazon.smithy.swift.codegen.utils
+
+import software.amazon.smithy.swift.codegen.SwiftSettings
+
+class ModelFileUtils {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/utils/ModelFileUtils.kt:5]: This is the utility / helper method used to select a filename for a model source file.

**Review by sichanyoo — APPROVED** [review]

**Pull request merged**